pypdf/filters.py

```diff
@@ -742,108 +742,97 @@ def _xobj_to_image(x_object_obj: Dict[str, Any]) -> Tuple[Optional[str], bytes,
         )
     extension = None
     alpha = None
-
-    if SA.FILTER in x_object_obj:
-        if x_object_obj[SA.FILTER] == FT.FLATE_DECODE or x_object_obj[SA.FILTER] == [
-            FT.FLATE_DECODE
-        ]:
-            extension = ".png"  # mime_type = "image/png"
-            if isinstance(color_space, ArrayObject) and color_space[0] == "/Indexed":
-                color_space, base, hival, lookup = (
-                    value.get_object() for value in color_space
-                )
-            img = Image.frombytes(mode, size, data)
-            if color_space == "/Indexed":
-                from .generic import ByteStringObject
-
-                if isinstance(lookup, ByteStringObject):
-                    try:
-                        nb, conv, mode = {  # type: ignore
-                            "1": (0, "", ""),
-                            "L": (1, "P", "L"),
-                            "P": (0, "", ""),
-                            "RGB": (3, "P", "RGB"),
-                            "CMYK": (4, "P", "CMYK"),
-                        }[_get_imagemode(base, 0, "")]
-                    except KeyError:  # pragma: no cover
+    filters = x_object_obj.get(SA.FILTER, [None])
+    lfilters = filters[-1] if isinstance(filters, list) else filters
+    if lfilters == FT.FLATE_DECODE:
```

Comment:
I think the code might be more readable if we had some extra functions. Something like this:

```python
if lfilters == FT.FLATE_DECODE:
    img, image_format, extension = _handle_flate(size, data, mode, color_space)
elif lfilters in (FT.LZW_DECODE, FT.ASCII_85_DECODE, FT.CCITT_FAX_DECODE):
    # Code block: Maybe leave that as-is
elif lfilters == FT.JPX_DECODE:
    img, image_format, extension = _handle_flate(size, data, mode, color_space)
elif lfilters == FT.CCITT_FAX_DECODE:
    # Code block: Maybe leave that also as is
elif lfilters is None:
    # Code block: Maybe leave that also as is
```